src/util/sandbox/fd_sandbox.c

```diff
@@ -206,140 +156,369 @@ switch_user( uint uid, uint gid ) {
+
+    /* Deprecated resources, not used. */
+    // { .resource=RLIMIT_LOCKS,      .limit=0UL             },
+    // { .resource=RLIMIT_RSS,        .limit=0UL             },
+  };
+
+  for( ulong i=0UL; i<sizeof(rlimits)/sizeof(rlimits[ 0 ]); i++ ) {
+    struct rlimit limit = { .rlim_cur=rlimits[ i ].limit, .rlim_max=rlimits[ i ].limit };
+    if( -1==setrlimit( rlimits[ i ].resource, &limit ) ) FD_LOG_ERR(( "setrlimit(%d) failed (%i-%s)", rlimits[ i ].resource, errno, fd_io_strerror( errno ) ));
   }
 }
 
-void *
-fd_sandbox_alloc_protected_pages( ulong page_cnt,
-                                  ulong guard_page_cnt ) {
-#define PAGE_SZ (4096UL)
-  void * pages = mmap( NULL, (2UL*guard_page_cnt+page_cnt)*PAGE_SZ, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0UL );
-  if( FD_UNLIKELY( pages==MAP_FAILED ) ) FD_LOG_ERR(( "mmap failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+void
+fd_sandbox_private_drop_caps( ulong cap_last_cap ) {
+  if( -1==prctl( PR_SET_SECUREBITS,
+                 SECBIT_KEEP_CAPS_LOCKED | SECBIT_NO_SETUID_FIXUP |
+                    SECBIT_NO_SETUID_FIXUP_LOCKED | SECBIT_NOROOT |
+                    SECBIT_NOROOT_LOCKED | SECBIT_NO_CAP_AMBIENT_RAISE |
+                    SECBIT_NO_CAP_AMBIENT_RAISE_LOCKED ) ) FD_LOG_ERR(( "prctl(PR_SET_SECUREBITS) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+
+  for( ulong cap=0UL; cap<=cap_last_cap; cap++ ) {
+    if( -1==prctl( PR_CAPBSET_DROP, cap, 0, 0, 0 ) ) FD_LOG_ERR(( "prctl(PR_CAPBSET_DROP) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  }
+
+  struct __user_cap_header_struct hdr = { _LINUX_CAPABILITY_VERSION_3, 0 };
+  struct __user_cap_data_struct   data[2] = { { 0 } };
+  if( -1==syscall( SYS_capset, &hdr, data ) )                          FD_LOG_ERR(( "syscall(SYS_capset) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  if( -1==prctl( PR_CAP_AMBIENT, PR_CAP_AMBIENT_CLEAR_ALL, 0, 0, 0 ) ) FD_LOG_ERR(( "prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_CLEAR_ALL) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+}
 
-  uchar * middle_pages = (uchar *)( (ulong)pages + guard_page_cnt*PAGE_SZ );
+struct landlock_ruleset_attr {
+    __u64 handled_access_fs;
+};
 
-  /* Make the guard pages untouchable */
-  if( FD_UNLIKELY( mprotect( pages, guard_page_cnt*PAGE_SZ, PROT_NONE ) ) )
-    FD_LOG_ERR(( "mprotect failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+void
+fd_sandbox_private_landlock_restrict_self( void ) {
+  struct landlock_ruleset_attr attr = {
+    .handled_access_fs = 0, /* No access to anything. */
+  };
 
-  if( FD_UNLIKELY( mprotect( middle_pages+page_cnt*PAGE_SZ, guard_page_cnt*PAGE_SZ, PROT_NONE ) ) )
-    FD_LOG_ERR(( "mprotect failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  long landlock_fd = syscall( SYS_landlock_create_ruleset, &attr, 8, 0 );
+  if( -1L==landlock_fd && (errno==ENOSYS || errno==EOPNOTSUPP ) ) {
+    FD_LOG_DEBUG(( "landlock_create_ruleset() not supported on this system" ));
```

Comment:
I'd make this a warning but clarify that it's a defense-in-depth measure. I think people telling us that it isn't supported on their system is a good thing